sonarts-core/src/rules/noAccessorFieldMismatchRule.ts

```diff
@@ -0,0 +1,139 @@
+    super.visitSetAccessor(accessor);
+  }
+
+  protected visitGetAccessor(accessor: ts.AccessorDeclaration): void {
+    this.visitAccessor(accessor, {type: "getter", name : Walker.getName(accessor)});
+    super.visitGetAccessor(accessor);
+  }
+
+  private visitAccessor(accessor: ts.MethodDeclaration | ts.AccessorDeclaration, setterOrGetter: {type: string, name: string} | undefined) {
+    if (!setterOrGetter) {
+      return;
+    }
+    const matchingField = Walker.matchingField(accessor, setterOrGetter.name);
+    if (Walker.isPublic(accessor) && accessor.body && matchingField && !this.fieldIsUsed(accessor, matchingField)) {
+      this.addFailureAtNode(accessor.name, `Refactor this ${setterOrGetter.type} so that it actually refers to the property '${matchingField.name.getText()}'`);
+    }
+  }
+
+  private static isPublic(method: ts.MethodDeclaration | ts.AccessorDeclaration) {
+    const modifier = nav.accessModifier(method);
+    return modifier && modifier.kind === ts.SyntaxKind.PublicKeyword;
+  }
+
+  private static getName(accessor: ts.AccessorDeclaration) {
+    return accessor.name.getText().toLowerCase();
+  }
+
+  private setterOrGetter(method: ts.MethodDeclaration | ts.AccessorDeclaration): {type: string, name: string} | undefined {
+    const name = method.name.getText().toLowerCase();
+    if (method.kind === ts.SyntaxKind.SetAccessor) {
+      return { type: "setter", name};
+    }
+    if (method.kind === ts.SyntaxKind.GetAccessor) {
+      return { type: "getter", name};
+    }
+    if (name.startsWith("set") || name.startsWith("Set")) {
+      if (method.parameters.length == 1) {
+        return { type: "setter", name: name.substring(3) };
+      }
+    }
+    if (name.startsWith("get") || name.startsWith("Get")) {
+      if (method.parameters.length == 0) {
+        return { type: "getter", name: name.substring(3)};
+      }
+    }
+  }
+
+  private static matchingField(method: ts.MethodDeclaration | ts.AccessorDeclaration, targetName: string): ts.PropertyDeclaration | ts.ParameterDeclaration | undefined {
+    const containingClass = nav.firstAncestor(method, [ts.SyntaxKind.ClassDeclaration, ts.SyntaxKind.ClassExpression]) as ts.ClassDeclaration | ts.ClassExpression;
+    if (!containingClass) return;
+    return containingClass.members
+      .filter(element => element.kind === ts.SyntaxKind.PropertyDeclaration)
+      .map(element => element as ts.PropertyDeclaration | ts.ParameterDeclaration)
+      .concat(Walker.fieldsDeclaredInConstructorParameters(containingClass))
+      .find(element => Walker.fieldNameMatches(element.name.getText(), targetName));
+  }
+
+  private static fieldNameMatches(fieldName: string, targetName: string): boolean {
+    const fieldNameLowerCase = fieldName.toLowerCase();
+    const underscoredTargetName = "_" + targetName;
+    return fieldNameLowerCase == targetName || fieldNameLowerCase == underscoredTargetName;
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') == should be === [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS1440)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') == should be === [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS1440)